flow/BlobCipher.cpp

```diff
@@ -112,25 +123,77 @@ BlobCipherKeyIdCache::BlobCipherKeyIdCache(EncryptCipherDomainId dId)
 	TraceEvent("Init_BlobCipherKeyIdCache").detail("DomainId", domainId);
 }
 
+BlobCipherKeyIdCacheKey BlobCipherKeyIdCache::getCacheKey(const EncryptCipherBaseKeyId& baseCipherKeyId,
+                                                          const EncryptCipherRandomSalt& salt) {
+	return std::make_pair(baseCipherKeyId, salt);
+}
+
 Reference<BlobCipherKey> BlobCipherKeyIdCache::getLatestCipherKey() {
-	return getCipherByBaseCipherId(latestBaseCipherKeyId);
+	return getCipherByBaseCipherId(latestBaseCipherKeyId, latestRandomSalt);
 }
 
-Reference<BlobCipherKey> BlobCipherKeyIdCache::getCipherByBaseCipherId(EncryptCipherBaseKeyId baseCipherKeyId) {
-	BlobCipherKeyIdCacheMapCItr itr = keyIdCache.find(baseCipherKeyId);
+Reference<BlobCipherKey> BlobCipherKeyIdCache::getCipherByBaseCipherId(const EncryptCipherBaseKeyId& baseCipherKeyId,
+                                                                       const EncryptCipherRandomSalt& salt) {
+	BlobCipherKeyIdCacheMapCItr itr = keyIdCache.find(getCacheKey(baseCipherKeyId, salt));
 	if (itr == keyIdCache.end()) {
+		TraceEvent("CipherByBaseCipherId_KeyMissing")
+		    .detail("DomainId", domainId)
+		    .detail("BaseCipherId", baseCipherKeyId)
+		    .detail("Salt", salt);
 		throw encrypt_key_not_found();
 	}
 	return itr->second;
 }
 
-void BlobCipherKeyIdCache::insertBaseCipherKey(EncryptCipherBaseKeyId baseCipherId,
+void BlobCipherKeyIdCache::insertBaseCipherKey(const EncryptCipherBaseKeyId& baseCipherId,
```

Comment:
btw since `EncryptCipherBaseKeyId` (an also `EncryptCipherDomainId` and `EncryptCipherRandomSatl`) is integer, shall we pass it by value? If you are fine with it, I can do it with my upcoming PR.

Reply:
I was passing it by value at first, however, I thought given we are using _alias_ I thought better to keep it as a reference, incase we update the field-type in future (silent failures or unexpected behavior); from perf point of view, either option should be equally in-expensive. 
Thanks